<launch>
    <!-- Configurable arguments -->
    <arg name="robot_name" default="girona500" />

    <arg name="enable_joystick" default="true" />
    <arg name="joystick_device" default="/dev/input/js0" />

    <arg name="enable_rviz" default="true" />
    <arg name="enable_gui" default="true" /> <!-- for iquaview -->

    <arg name="gripper_ns" default="girona500/bravo/gripper"/>
    <arg name="nodelet_manager" default="nodelet_manager"/>
    <arg name="respawn" default="false"/>
    <arg name="bond" default="true"/>
    <!-- Simulation parameters -->
    <arg name="simulation_data" value="$(find tandem_stonefish)/resources" />
    <arg name="scenario_description"
        value="$(find tandem_stonefish)/scenarios/girona500_uji_simulation.scn" />
    <arg name="simulation_rate" value="300.0" />
    <arg name="graphics_resolution" value="1200 800" />
    <arg name="graphics_quality" value="high" />

    <!-- Static transform -->
    <node name="world2ned" pkg="tf2_ros" type="static_transform_publisher"
        args="0 0 0 3.1415 0 3.1415 world world_ned" />
        
    <include file="$(find girona500_description)/launch/cirtesu_description.launch"/>

    <!-- Namespace with robot name -->
    <group ns="$(arg robot_name)">

        <!-- Architecture -->
        <include file="$(find tandem_stonefish)/launch/core/girona500_uji_base.launch.core">
            <arg name="robot_name" value="$(arg robot_name)" />
            <arg name="enable_gui" value="$(arg enable_gui)" />
        </include>

        <!-- Joystick -->
        <group if="$(arg enable_joystick)">
            <node name="joystick" pkg="joy" type="joy_node">
                <param name="dev" value="$(arg joystick_device)" />
                <param name="dev_ff" value="" />
                <param name="deadzone" value="0.05" />
            </node>
            <rosparam ns="logitech_fx10_to_teleoperation" command="load"
                file="$(find tandem_stonefish)/config/logitech_fx10_to_teleoperation.yaml" />
            <node name="logitech_fx10_to_teleoperation" pkg="tandem_stonefish"
                type="joy_to_teleoperation.py" output="screen">
                <param name="bravo_ns" value="bravo" />
            </node>

            <node name="body_velocity_republisher" pkg="intervention_cirtesu" type="tp_body_vel_relay.py" output="screen">
                <param name="sub_topic" value="/girona500/plane_inspection/cmd_vel"/>
                <param name="pub_topic" value="/girona500/controller/body_velocity_req"/>
                <param name="service_name" value="/girona500/tp_controller/active"/>
            </node>        
            <!-- <node name="fx10_to_teleoperation" pkg="cola2_control"
                type="logitech_fx10_to_teleoperation_node.py" /> -->
        </group>

        <!-- Load robot description -->
        <param name="robot_description"
            command="$(find xacro)/xacro '$(find tandem_stonefish)/urdf/girona500_uji_bravo_payload_real.urdf.xacro'
                        robot_namespace:=$(arg robot_name)" />
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
            output="screen" />

        <!-- Run Simulator -->
        <node name="stonefish_simulator" pkg="stonefish_ros" type="parsed_simulator"
            args="$(arg simulation_data) $(arg scenario_description)
                  $(arg simulation_rate) $(arg graphics_resolution) 
                  $(arg graphics_quality)"
            output="screen">
            <remap from="bravo/joint_states" to="joint_states" />
        </node>

        <!-- ROS Control -->
        <group ns="bravo">
            <!-- Load ros-control parameters -->
            <rosparam command="load"
                file="$(find tandem_stonefish)/config/ros_control/bravo_uji_ros_control.yaml" />
            <!-- Load the controllers -->
            <node name="controller_spawner" pkg="controller_manager" type="spawner" output="screen"
                args="joint_state_controller" />
            
            <node name="load_controllers" pkg="controller_manager" type="controller_manager" respawn="false" output="screen" 
                args="load joint_teleop_velocity_controller joint_trajectory_controller 
                joint_velocity_controller gripper_velocity_controller_finger_small 
                gripper_velocity_controller_finger_large gripper_position_controller_finger_small gripper_position_controller_finger_large"/>

            <!-- Launch the predefined motions node-->
            <node name="predefined_positions" pkg="tandem_stonefish"
                type="predefined_positions" output="screen">
                <rosparam command="load"
                    file="$(find tandem_stonefish)/config/predefined_motions/bravo_uji.yaml" />
            </node>
            <!-- <group if="$(arg enable_joystick)">
                <node respawn="true" pkg="joy" type="joy_node" name="joystick">
                    <param name="dev" value="/dev/input/js0" />
                    <param name="dev_ff" value="" />
                    <param name="deadzone" value="0.05" />
                </node>
                
                <rosparam ns="logitech_fx10_to_teleoperation" command="load"
                    file="$(find tandem_stonefish)/config/logitech_fx10_to_teleoperation.yaml" />

                <node name="logitech_fx10_to_teleoperation" pkg="tandem_stonefish"
                    type="joy_to_teleoperation.py" output="screen">
                    <param name="bravo_ns" value="bravo" />
                </node>
            </group> -->
        </group>

    </group>

    <!-- dummy node to fold the manipulators -->
    <!-- <node name="start_arms_folded" pkg="tandem_stonefish" type="fold_arms.sh" /> -->


    <!-- Stonefish/COLA2 message translator -->
    <node name="translator" pkg="cola2_stonefish" type="cola2_stonefish_node">
        <remap from="/translator/stonefish_dvl" to="/$(arg robot_name)/navigator/dvl_sim" />
        <remap from="/translator/cola2_dvl" to="/$(arg robot_name)/navigator/dvl" />
        <remap from="/translator/stonefish_thruster_setpoints"
            to="/$(arg robot_name)/controller/thruster_setpoints_sim" />
        <remap from="/translator/cola2_thruster_setpoints"
            to="/$(arg robot_name)/controller/thruster_setpoints" />
        <remap from="/translator/stonefish_navigation"
            to="/$(arg robot_name)/navigator/navigation_sim" />
        <remap from="/translator/cola2_navigation" to="/$(arg robot_name)/navigator/navigation" />
    </node>

    <!-- RViz -->
    <node if="$(arg enable_rviz)" pkg="rviz" type="rviz" name="rviz"
        args="-d $(find tandem_stonefish)/config/rviz/girona500_uji_real_cameras.rviz" />

    <!-- Rqt-->
    <node name="rqt_gui" pkg="rqt_gui" type="rqt_gui">
        <remap from="robot_description" to="/girona500/robot_description"/>
    </node>
    <!-- Aruco detector-->
    <group ns="girona500/bravo/gripper">
        <node name="$(arg nodelet_manager)" pkg="nodelet" type="nodelet" args="manager" output="screen" />
        <node name="aruco_tracker" pkg="nodelet" type="nodelet"
            args="load aruco_opencv/ArucoTracker /$(arg gripper_ns)/$(arg nodelet_manager) $(arg bond)" 
            output="screen" respawn="$(arg respawn)">
            <rosparam command="load"
            file="$(find tandem_stonefish)/config/aruco/aruco_tracker.yaml" />
            <param name="board_descriptions_path"
            value="$(find tandem_stonefish)/config/aruco/board_descriptions.yaml" />
        </node>
    </group>
</launch>
